pkg/grpc/remote_authenticator.go

```diff
@@ -0,0 +1,181 @@
+// remoteAuthenticator.
+type RemoteAuthenticatorCacheKey [sha256.Size]byte
+
+type remoteAuthCacheEntry struct {
+	ready    <-chan struct{}
+	response remoteAuthResponse
+}
+
+type remoteAuthResponse struct {
+	expirationTime time.Time
+	authMetadata   *auth.AuthenticationMetadata
+	err            error
+}
+
+func (ce *remoteAuthCacheEntry) HasExpired(now time.Time) bool {
+	select {
+	case <-ce.ready:
+		return ce.response.expirationTime.Before(now)
+	default:
+		// Ongoing remote requests have not expired by definition.
+		return false
+	}
+}
+
+// NewRemoteAuthenticator creates a new RemoteAuthenticator for incoming
+// requests that forwards headers to a remote service for authentication. The
+// result from the remote service is cached.
+func NewRemoteAuthenticator(
+	client grpc.ClientConnInterface,
+	scope *structpb.Value,
+	clock clock.Clock,
+	evictionSet eviction.Set[RemoteAuthenticatorCacheKey],
+	maximumCacheSize int,
+) RequestHeadersAuthenticator {
+	return &remoteAuthenticator{
+		remoteAuthClient: auth_pb.NewAuthenticationClient(client),
+		scope:            scope,
+
+		clock:            clock,
+		maximumCacheSize: maximumCacheSize,
+
+		cachedResponses: make(map[RemoteAuthenticatorCacheKey]*remoteAuthCacheEntry),
+		evictionSet:     evictionSet,
+	}
+}
+
+func (a *remoteAuthenticator) Authenticate(ctx context.Context, headers map[string][]string) (*auth.AuthenticationMetadata, error) {
+	request := &auth_pb.AuthenticateRequest{
+		RequestMetadata: make(map[string]*auth_pb.AuthenticateRequest_ValueList, len(headers)),
+		Scope:           a.scope,
+	}
+	for headerKey, headerValues := range headers {
+		request.RequestMetadata[headerKey] = &auth_pb.AuthenticateRequest_ValueList{
+			Value: headerValues,
+		}
+	}
+	requestBytes, err := proto.Marshal(request)
+	if err != nil {
+		return nil, util.StatusWrapWithCode(err, codes.Unauthenticated, "Failed to marshal authenticate request")
+	}
+	// Hash the request to use as a cache key to both save memory and avoid
+	// keeping credentials in the memory.
+	requestKey := sha256.Sum256(requestBytes)
+
+	a.lock.Lock()
+	now := a.clock.Now()
+	entry := a.getAndTouchCacheEntry(requestKey)
+	if entry != nil && entry.HasExpired(now) {
+		entry = nil
+	}
+	if entry == nil {
+		// No valid cache entry available. Deduplicate requests by creating a
+		// pending cached response.
+		responseReady := make(chan struct{})
+		entry = &remoteAuthCacheEntry{
+			ready: responseReady,
+		}
+		a.cachedResponses[requestKey] = entry
+		a.lock.Unlock()
+
+		// Perform the remote authentication request.
+		entry.response = a.authenticateRemotely(ctx, request)
+		close(responseReady)
+	} else {
+		a.lock.Unlock()
+
+		// Wait for the remote request to finish.
+		select {
+		case <-ctx.Done():
+			return nil, util.StatusWrapWithCode(ctx.Err(), codes.Unauthenticated, "Context cancelled")
+		case <-entry.ready:
+			// Noop
+		}
+	}
+	return entry.response.authMetadata, entry.response.err
```

Comment:
I don't think this is sound. What happens if `entry.response.err` is of type `codes.Canceled`? We end up returning a cancelation error belonging to a different client.

Reply:
Would it be correct to use `contect.Background()` in the request?

Reply:
That could work, but has the downside that cancelation isn't propagated immediately. My recommendation would be to stick to `ctx`, but don't coalesce error results of in-flight RPCs.